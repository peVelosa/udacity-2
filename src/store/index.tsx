import { combineReducers, configureStore } from "@reduxjs/toolkit";
import { authSlice } from "./auth/auth.reducer";
import { poolSlice } from "./pool/pool.reducer";
import { usersSlice } from "./users/users.reducer";

export const rootReducer = combineReducers({
  auth: authSlice.reducer,
  questions: poolSlice.reducer,
  users: usersSlice.reducer,
});

export function setupStore(preloadedState?: Partial<RootState>) {
  return configureStore({
    reducer: rootReducer,
    preloadedState,
  });
}
export type RootState = ReturnType<typeof rootReducer>;
export type AppStore = ReturnType<typeof setupStore>;
export type AppDispatch = AppStore["dispatch"];
