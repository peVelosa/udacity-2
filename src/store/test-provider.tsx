import React, { PropsWithChildren } from "react";
import { render } from "@testing-library/react";
import type { RenderOptions } from "@testing-library/react";
import { Provider } from "react-redux";
import { RootState, AppStore, setupStore } from ".";
import { MemoryRouter, Routes, Route, InitialEntry } from "react-router";

interface ExtendedRenderOptions extends Omit<RenderOptions, "queries"> {
  preloadedState?: Partial<RootState>;
  store?: AppStore;
}

export function renderWithProviders(
  ui: React.ReactElement,
  extendedRenderOptions: ExtendedRenderOptions = {},
  initialEntries: InitialEntry[],
  path: string
) {
  const {
    preloadedState = {},
    store = setupStore(preloadedState),
    ...renderOptions
  } = extendedRenderOptions;

  const Wrapper = ({ children }: PropsWithChildren) => (
    <Provider store={store}>
      <MemoryRouter initialEntries={initialEntries}>
        <Routes>
          <Route path={path} element={children} />
        </Routes>
      </MemoryRouter>
      ,
    </Provider>
  );

  return {
    store,
    ...render(ui, { wrapper: Wrapper, ...renderOptions }),
  };
}
