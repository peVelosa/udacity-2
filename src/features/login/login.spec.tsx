import "@testing-library/jest-dom";
import { LoginPage } from ".";
import { renderWithProviders } from "@/store/test-provider";
import { fireEvent, waitFor } from "@testing-library/react";
import * as server from "@/server";

describe("LoginPage", () => {
  test("should render the login page", () => {
    const { getByTestId } = renderWithProviders(<LoginPage />);
    const user = getByTestId("user-field");
    const password = getByTestId("password-field");
    const button = getByTestId("submit-button");

    expect(user).toBeInTheDocument();
    expect(password).toBeInTheDocument();
    expect(button).toBeInTheDocument();
  });
  test("shows error on login with invalid credentials", async () => {
    jest.spyOn(server, "_getUsers").mockResolvedValue({
      tylermcginnis: { password: "abc321" },
    } as any);

    const alertSpy = jest.spyOn(window, "alert").mockImplementation(() => {});

    const { getByTestId } = renderWithProviders(<LoginPage />);
    const userField = getByTestId("user-field");
    const passwordField = getByTestId("password-field");
    const submitButton = getByTestId("submit-button");

    fireEvent.change(userField, { target: { value: "wronguser" } });
    fireEvent.change(passwordField, { target: { value: "wrongpass" } });

    fireEvent.click(submitButton);

    await waitFor(() => {
      expect(alertSpy).toHaveBeenCalledWith("User not found");
    });

    alertSpy.mockRestore();
  });
  test("login button should be disabled with any empty field", () => {
    const { getByTestId } = renderWithProviders(<LoginPage />);
    const user = getByTestId("user-field");
    const password = getByTestId("password-field");
    const button = getByTestId("submit-button");

    expect(button).toBeInTheDocument();
    expect(button).toBeDisabled();

    fireEvent.change(user, { target: { value: "tylermcginnis" } });
    expect(button).toBeDisabled();

    fireEvent.change(password, { target: { value: "wrongpass" } });
    expect(button).toBeEnabled();
  });
});
