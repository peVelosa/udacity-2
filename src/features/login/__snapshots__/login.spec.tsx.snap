// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginPage matches snapshot 1`] = `
<DocumentFragment>
  <section
    class="mt-20"
  >
    <h1
      class="text-center text-xl font-bold"
    >
      Log In
    </h1>
    <form
      class="flex flex-col gap-4 text-center"
    >
      <div>
        <label
          class="block"
          for="user"
        >
          User
        </label>
        <input
          class="w-full border border-neutral-300 rounded-sm p-2"
          data-testid="user-field"
          id="user"
          placeholder="User"
          testid="user-field"
          value=""
        />
      </div>
      <div>
        <label
          class="block"
          for="password"
        >
          Password
        </label>
        <input
          class="w-full border border-neutral-300 rounded-sm p-2"
          data-testid="password-field"
          id="password"
          placeholder="Password"
          testid="password-field"
          type="password"
          value=""
        />
      </div>
      <button
        class="px-4 py-2 bg-neutral-900 w-fit text-white font-bold rounded-md mx-auto disabled:bg-neutral-500"
        data-testid="submit-button"
        disabled=""
        type="submit"
      >
        Submit
      </button>
    </form>
  </section>
  ,
</DocumentFragment>
`;
